src/crab/ebpf_domain.cpp

```diff
@@ -673,11 +705,98 @@ void ebpf_domain_t::operator()(const Un& stmt) {
     auto dst = reg_pack(stmt.dst);
     switch (stmt.op) {
     case Un::Op::BE16:
+        if (m_inv.entail(type_is_number(stmt.dst))) {
```

Comment:
The repetition here can be avoided (though maybe too cleverly) using a template lambda. Something like this:
```cpp
void ebpf_domain_t::operator()(const Un& stmt) {
    auto dst = reg_pack(stmt.dst);
    auto swap_endianness = [&](auto input, const auto& be_or_le) -> void {
        // input serves as a type parameter; explicit type parameters is a c++20 feature
        if (m_inv.entail(type_is_number(stmt.dst))) {
            auto interval = m_inv.eval_interval(dst.value);
            if (std::optional<number_t> n = interval.singleton()) {
                if (n->fits_uint64()) {
                    input = (decltype(input))(uint64_t)n.value();
                    number_t output{(uint32_t)be_or_le(input)}; // uint32_t to avoid ambiguity
                    const crab::interval_t v{output, output};
                    m_inv.set(dst.value, v);
                    return;
                }
            }
        }
        havoc(dst.value);
        havoc_offsets(stmt.dst);
    };
    switch (stmt.op) {
    case Un::Op::BE16:
        swap_endianness(uint16_t(0), boost::endian::native_to_big<uint16_t>);
        break;
    case Un::Op::BE32:
        swap_endianness(uint32_t(0), boost::endian::native_to_big<uint32_t>);
        break;
    case Un::Op::BE64:
        swap_endianness(uint64_t(0), boost::endian::native_to_big<uint64_t>);
        break;
    case Un::Op::LE16:
        swap_endianness(uint16_t(0), boost::endian::native_to_little<uint16_t>);
        break;
    case Un::Op::LE32:
        swap_endianness(uint32_t(0), boost::endian::native_to_little<uint32_t>);
        break;
    case Un::Op::LE64:
        swap_endianness(uint64_t(0), boost::endian::native_to_little<uint64_t>);
        break;
    case Un::Op::NEG:
        neg(dst.value);
        havoc_offsets(stmt.dst);
        break;
    }
}
```
A template member function is also possible.

Reply:
Done (converted to a templated member function)